teps/0091-trusted-resources.md

```diff
@@ -313,6 +395,102 @@ What other approaches did you consider and why did you rule them out?  These do
+
+### Options for storing the public keys
+
+Option 1: Configure the keys in a dedicated CRD. This can help to support multiple keys in a central place.
+
+Option 2: Add a new field `keyref` to refer to the key path
+```yaml
+kind: TaskRun
+apiVersion: tekton.dev/v1beta1
+metadata:
+ name: example-tr
+spec:
+ taskRef:
+   name: example-task
+   keyref: keypath # this defines the public key of this taskref
+```
+
+```yaml
+kind: PipelineRun
+apiVersion: tekton.dev/v1beta1
+metadata:
+ name: example-pr
+spec:
+ pipelineRef:
+   name: example-pipeline
+   keyref: keypath # this defines the public key of this pipelineref
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+ annotations:
+   tekton.dev/signature: MEQCIHhFC480mv7mh/6dxAlp/mvuvGXanuSghMsT+iBhWxt5AiBDvHv8sfKjJ3Ozrzvp+cjr30AOx1SPQDCcaJjlpAoVrA==
+ name: example-Pipeline
+spec:
+ tasks:
+   - name: example-task
+     taskRef:
+       name: example-task1
+       keyref: keypath1  # this defines the public key of this taskref
+   - name: example-task2
+     taskRef:
+       name: example-task2
+       keyref: keypath2  # this defines the public key of this taskref
+```
+`keyref` defines the path of the public key of this taskRef. All the tasks within a pipeline need to be signed off, and they may come from different sources with different public keys.
+So a `keyref` is needed when signing the pipeline. The signing cli doesn't need to verify the referred keys when signing the pipeline.
+
+**Pros vs Cons:**
+| Option | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Config at install level (e.g. at CRD) | Easy to config at a central place | Need to loop all the keys for each verification
+| Add keyref in taskRef&pipelineRef | Each resource can specify the corresponding key, no need to loop all available keys for one resource's verification| Introduce cost for users to write and maintain YAML files
+
+### Options for integrating with remote resolution
+
+| Method | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Fetch remote resources in validating Webhook | Verification fail fast in webhook | Duplicate work for Resolution, may introduce latency, and one extreme case is that the resource verified may not the the same in reconciler's resolved resource
+| Verify in Controller after resolving resources| No duplicate work for Resolution | The verification cannot fail fast in webhook. The resources may have been stored in etcd and used by other components
+| Verify in Remote Resolution | No duplicate work for Resolution | Verification coupled with Resolution
+
+In this TEP we propose to fetch remote resources in validating Webhook considering that untrusted resources shouldn't be cond in etcd. And the performance influence of duplicate resolution can be ignored.
```

Comment:
The way remote resolution works, you'll have to create a resolution request crd from the webhook, wait for the resolver to do its thing and update the status of the resolution request with the pipeline content and then verify. Essentially you'll be implementing a reconciler in the webhook itself :) I don't think we should do that. 

What is the issue with creating a run object that contains a reference to a potentially untrusted resource and storing it in etcd?

Reply:
Oh sorry I forgot to update this part. We have reached to an agreement to do it in the reconciler and not in the webhook. So this won't be a problem.